lib/integrations/awsoidc/deployservice.go

```diff
@@ -0,0 +1,696 @@
+	serviceStatusDraining = "DRAINING"
+)
+
+var (
+	// DatabaseServiceDeploymentMode is a deployment configuration for Deploying a Database Service.
+	// This mode starts a Database with the specificied Resource Matchers.
+	DatabaseServiceDeploymentMode = "database-service"
+
+	// DeploymentModes has all the available deployment modes.
+	DeploymentModes = []string{
+		DatabaseServiceDeploymentMode,
+	}
+)
+
+// DeployServiceRequest contains the required fields to deploy a Teleport Service.
+type DeployServiceRequest struct {
+	// Region is the AWS Region
+	Region string
+
+	// SubnetIDs are the subnets associated with the service.
+	SubnetIDs []string
+
+	// ClusterName is the ECS Cluster to be used.
+	// It will be created if it doesn't exist.
+	// It will be updated if it doesn't include the FARGATE capacity provider using PutClusterCapacityProviders.
+	ClusterName *string
+
+	// ServiceName is the ECS Service to be used.
+	// It will be created if it doesn't exist.
+	// It will be updated if it doesn't match the required properties.
+	ServiceName *string
+
+	// TaskName is the ECS Task Definition's name.
+	TaskName *string
+
+	// TaskRoleARN is the AWS Role's ARN used within the Task execution.
+	// Ensure the AWS Client has `iam:PassRole` for this Role's ARN.
+	TaskRoleARN string
+
+	// TeleportClusterName is the Teleport Cluster Name, used to create default names for Cluster, Service and Task.
+	TeleportClusterName string
+
+	// ProxyServerHostPort is the Teleport Proxy's Public.
+	ProxyServerHostPort string
+
+	// IntegrationName is the integration name.
+	// Used for resource tagging when creating resources in AWS.
+	IntegrationName string
+
+	// ResourceCreationTags is used to add tags when creating resources in AWS.
+	ResourceCreationTags awsTags
+
+	// DeploymentMode is the identifier of a deployment mode - which Teleport Services to enable and their configuration.
+	DeploymentMode string
+
+	// DatabaseResourceMatcherLabels contains the set of labels to be used by the DatabaseService.
+	// This is used when the deployment mode creates a Database Service.
+	DatabaseResourceMatcherLabels types.Labels
+}
+
+// CheckAndSetDefaults checks if the required fields are present.
+func (r *DeployServiceRequest) CheckAndSetDefaults() error {
+	if r.TeleportClusterName == "" {
+		return trace.BadParameter("teleport cluster name is required")
+	}
+
+	if r.Region == "" {
+		return trace.BadParameter("region is required")
+	}
+
+	if len(r.SubnetIDs) == 0 {
+		return trace.BadParameter("at least one subnet id is required")
+	}
+
+	if r.TaskRoleARN == "" {
+		return trace.BadParameter("task role arn is required")
+	}
+
+	if r.ClusterName == nil || *r.ClusterName == "" {
+		clusterName := fmt.Sprintf("%s-teleport", r.TeleportClusterName)
+		r.ClusterName = &clusterName
+	}
+
+	if r.ServiceName == nil || *r.ServiceName == "" {
+		serviceName := fmt.Sprintf("%s-teleport-service", r.TeleportClusterName)
```

Comment:
Are there any limitations to the service/task name? I.e. if cluster name contains a dot is that ok?